### ISSUE TITLE
[INTENT TO DEPRECATE] -- @plugin

### ISSUE THREAD

**Issue by matthew-dean** [less/less-meta#36]
## TL;DR

**Instead of this:**
```less
@plugin "functions.js";

.rule {
  value: myFunction();
}
```
```js
// functions.js -- currently, `functions` is an injected var
functions.add('myFunction', function () {
  return new SomeNodeType();
});
```
**Future Less syntax would be this:**
```less
@import { myFunction } from "functions.js";

.rule {
  value: myFunction();
}
```
```js
// functions.js
export myFunction () {
  return new SomeNodeType();
}
```

## Some history

Historically, plugins were loaded before parse / render steps.

Later, a feature was submitted to add `@plugin` inline. However, that just added an eval-time function registry, so that functions could be added to scope. So, all of the sudden, Less had two types of JavaScript plugins, neither of which was compatible with the other.

To potentially solve that (later than that) the different "plugin" types were unified in one loading mechanism, such that `@plugin "some-plugin.js"` could now load the originally-defined type of plugin, complete with adding visitors, etc. This, though, created a lot of hacks, and made some visitor-types counter-intuitive.

To put it another way, we solved one problem. Now we had two problems.
 
Namely: if you add visitors during tree eval, you could add a visitor that visits a type of node that has already been visited. Also, functions added by `@plugin` were local to the local scope, but visitors were always global.

In short, `@plugin` is kind of a mess.

## Reasons to use `@import`

1. It restores the concept of "plugin" to a pre-parse/-render application. That is, a Less plugin is always global.
2. It borrows a known concept (ES2015-style imports).
3. It makes function importing more declarative, and can be statically analyzed. There are no magic scoped variables. So it also makes JavaScript imports lint-able. 
4. We can potentially use Rollup to resolve any nested imports, rather than the injected `require` hack that is in `@plugin`.

### Future expansion

_**Not part of this proposal,**_ but expanding `@import` syntax would potentially allow future expansion such as:

```less
// main.less -- can now be statically analyzed to find the source of `@var`
@import { @var } from './variables.less';
.rule {
  value: @var;
}

// replacement for `@import (reference)`
@import * as #lib from './library.less';

#lib.mixin();

// pull just one mixin or rule
@import { .mixin } from './mixins.less';

.rule-2 {
  value: .mixin[@lookup];
}
```

**Comment by matthew-dean** [comment]
I've decided this pattern is a departure enough to move it to a new language: https://github.com/jesscss/jess
